### ISSUE TITLE
Narrowed types from user defined type guards are lost in subsequent method calls

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#20186]
**TypeScript Version:**  2.6.1.

**Code**

```ts
{
    // no error
    const arr1: Array<number> = ['foo', 1].filter(
        (x): x is number => typeof x === 'number',
    );

    /* unexpected error:
[ts]
Type '(string | number)[]' is not assignable to type 'number[]'.
Type 'string | number' is not assignable to type 'number'.
Type 'string' is not assignable to type 'number'.
*/
    const arr2: Array<number> = ['foo', 1]
        .filter((x): x is number => typeof x === 'number')
        .map(x => x);

    // workaround. no error
    const arr3: Array<number> = ['foo', 1]
        .filter((x: any): x is number => typeof x === 'number')
        .map(x => x);
}
```

Related https://github.com/ReactiveX/IxJS/issues/162

**Comment by trxcllnt** [comment]
possibly related to #19334. defining the type-guard before the filter call works fine:
```typescript
function isNumber(x: string | number): x is number {
  return typeof x === 'number';
}
// no error
const arr: Array<number> = ['foo', 1].filter(isNumber).map((x) => x);
```

**Comment by trxcllnt** [comment]
@OliverJAsh based on https://github.com/Microsoft/TypeScript/issues/19640#issuecomment-341122227 it looks like this may be fixed by #17600, as this form also works in 2.6.1:

```typescript
// no error
const arr: Array<number> = ['foo', 1]
  .filter<number>((x): x is number => typeof x === 'number')
  .map((x) => x);
```

**Comment by ahejlsberg** [comment]
This is indeed fixed by #17600.
